js-packages/@prestojs/viewmodel/src/useViewModelChangeEffect.ts

```diff
@@ -0,0 +1,174 @@
+// @flow
+import { useEffect, useRef } from 'react';
+import ViewModel from './ViewModel';
+
+export const ADD = 'ADD';
+export const UPDATE = 'UPDATE';
+export const DELETE = 'DELETE';
+
+type ChangeType = 'ADD' | 'UPDATE' | 'DELETE';
+
+type Options = {
+    /**
+     * If true then when any records are added it will call `run` with 'ADD'
+     * as the first parameter.
+     */
+    runOnAdd?: boolean;
+    /**
+     * If true then when any records are deleted it will call `run` with 'DELETE'
+     * as the first parameter.
+     */
+    runOnDelete?: boolean;
+    /**
+     * If true then when any records are updated it will call `run` with 'UPDATE'
+     * as the first parameter.
+     */
+    runOnUpdate?: boolean;
+    /** If true will pass through diff object to `run` on second parameter */
+    includeDiff?: boolean;
+};
+
+type AddedDiff<T extends ViewModel> = {
+    previous: T[];
+    added: T[];
+};
+
+type UpdatedDiff<T extends ViewModel> = {
+    previous: T[];
+    updated: T[];
+};
+
+type DeletedDiff<T extends ViewModel> = {
+    previous: T[];
+    deleted: T[];
```

Comment:
Is it going to be annoying to use if added/updated/deleted are different?

If the fields are the same `previousState` / `newState` regardless of whether it's an Added/Updated/Deleted operation then you don't need a `switch`/`if` based on the event type -- or is that something you're actively wanting to be treated differently?

(or the "light" version of this: just Added/Updated be the same thing since they're often treated the same way)

Reply:
they have to be different as they can all happen (the diff is a union of all these types - see `Diff` below)